### PR TITLE
[google_maps_flutter] iOS Pattern Support for Polylines

### PR DESCRIPTION
**Pull request by vanlooverenkoen**
Added the iOS implementation for pattern support with polylines.

I have a Java/Kotlin background so this was a real challenge for me. Please update the code if it should be implemented differently. But this is working fine for me. (I don't have experience with tests in iOS so please help me with that)

Updated the example as well

*List which issues are fixed by this PR. You must list at least one issue.*
Fixed: https://github.com/flutter/flutter/issues/60083

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [X] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [X] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [X] I updated CHANGELOG.md to add a description of the change.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [ ] All existing and new tests are passing.

### CODE REVIEW

**Comment by vanlooverenkoen** [comment]
How do I fix these `This user does not have permission to run Firebase Test Lab tests.` errors?

**Comment by vanlooverenkoen** [comment]
@ditman can you review this pull request as well?

**Comment by ditman** [comment]
> How do I fix these This user does not have permission to run Firebase Test Lab tests. errors?

@vanlooverenkoen when was this last rebased with `master`? I haven't seen those errors recently, maybe this branch is missing some config? (@cyanglaz might know more)

**Comment by vanlooverenkoen** [comment]
Last rebase was on Saturday.

**Comment by cyanglaz** [comment]
I've never seen this. try re-running

**Comment by stuartmorgan** [comment]
> I've never seen this. try re-running

It's persistent, I'm trying to figure out why it's happening offline. It doesn't need to block things for now though.

**Comment by stuartmorgan** [comment]
Thanks for the submission! We’re currently working through a large backlog of PRs, and this will require non-trivial review, so it will take some time before we’re able to review it. As explained in [CONTRIBUTING.md](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#the-review-process), votes for the corresponding issue are the primary way we’re prioritizing non-trivial reviews, so we encourage anyone interested in this PR to vote for the corresponding issue.

**Comment by vanlooverenkoen** [comment]
@cyanglaz I fixed my pull rewuesy

**Comment by vanlooverenkoen** [comment]
@cyanglaz & @ditman can you review my pull request. all checks have passed after merging from master.

**Comment by allanwolski** [comment]
I need this too.

**Comment by vincepunkrock** [comment]
Please reviewers, review and approve this PR :)

**Comment by chandaniBhalala** [comment]
Hello, i need dash polyline in iOS also. Can you suggest me how can i get this issue fix in urgent bases.

**Comment by stuartmorgan** [comment]
Please see my comment above about how reviews are prioritized.

Also, to save time once it reaches the review stage: the first thing a reviewer would currently say is that this PR [can't process without tests](https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#welcome). Implementing tests now will save significant time once the review starts.

**Comment by vanlooverenkoen** [comment]
I am not an iOS dev. So I have no idea how I should start with the integration tests. If someone could help me with that. That would be nice. Otherwise this PR that we are using in production for 6 months now will never be merged I guess.

**Comment by vanlooverenkoen** [comment]
@stuartmorgan. Is there somebody who can help me with these tests?

**Comment by stuartmorgan** [comment]
> @stuartmorgan. Is there somebody who can help me with these tests?

When it reaches review whoever is reviewing it can help with writing tests, yes. (Also, anyone interested in this PR should feel free to help with tests.)

This will likely need XCTests, which unfortunately I can't point to similar examples for in this plugin as there's a lot of missing test coverage right now.

**Comment by vanlooverenkoen** [comment]
Yes indeed. If anyone wants to help with this please let me know. 

This needs to be merged because it is a missing feature on iOS that works perfectly on android. We are using this in production for over 5 months now.

**Comment by stuartmorgan** [comment]
> This needs to be merged

Nobody has suggested that it won't be merged, but it also needs to follow our process and requirements.

> We are using this in production for over 5 months now.

And tests will ensure that everyone will be able to use it in their production apps, without it regressing later. The difference between your branch and the actual repository is that you are presumably not accepting lots patches from a variety of people making their own changes to the plugin, any of which could potentially break existing functionality.

**Comment by vanlooverenkoen** [comment]
Yes of course. But it is quite hard to find the right people to get me started with the tests :D Thanks anyway for your review already.

**Comment by junwha0511** [comment]
Is there nobody who can write the XCTest for him?
Many people are waiting for this branch to merge.

**Comment by vanlooverenkoen** [comment]
Anybody?

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/JsonConversions.h
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/ios/Classes/JsonConversions.h:8]: This class and its properties and methods need comments, per the style guide. (The existing code in this class unfortunately doesn't follow style, but that shouldn't be perpetuated to new code.)

File: packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapPolylineController.m
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter/ios/Classes/GoogleMapPolylineController.m:58]: These can be pre-created with size equal to `_pattern`'s size to avoid unnecessary reallocations.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Can you elaborate on what help specifically you are looking for on the testing aspect?Unit testing seems like it should be relatively straightforward, but I haven't looked in detail so maybe I'm missing something:
- Tests for the new `FLTPolylinePattern`
- Tests that calling `setPattern` populates `_polyline.spans` correctly

**Comment by vanlooverenkoen** [comment]
> Can you elaborate on what help specifically you are looking for on the testing aspect? Unit testing seems like it should be relatively straightforward, but I haven't looked in detail so maybe I'm missing something:
> 
> * Tests for the new `FLTPolylinePattern`
> * Tests that calling `setPattern` populates `_polyline.spans` correctly

I have no Swift or Objective C experience. I looked at the native Android code and tried to figure out, by trial and error to get the implementation correct.

**Comment by stuartmorgan** [comment]
Have you looked at tests for other plugins? https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_wkwebview/example/ios/RunnerTests/FLTCookieManagerTests.m for instance demonstrates some relatively straightforward unit tests.

**Comment by zzimickQ** [comment]
Hey, guys this PR has been open for 14 months. can someone please just fix the comments and merge it?

**Comment by stuartmorgan** [comment]
@zzimickQ You are welcome to contribute the necessary tests.

**Pull request closed without merging**
